[GStreamer] Video element keeps changing the aspect ratio randomly (when the orientation information is in video's metadata)

philn opened the pull request:
#### 07f6e1de7e305114b2f4aaeef0fc25459ad9b4d8
<pre>
[GStreamer] Video element keeps changing the aspect ratio randomly (when the orientation information is in video&apos;s metadata)
<a href="https://bugs.webkit.org/show_bug.cgi?id=246824">https://bugs.webkit.org/show_bug.cgi?id=246824</a>
&lt;rdar://problem/101646324&gt;

Reviewed by Xabier Rodriguez-Calvar.

Emit sizeChanged notifications towards the parent player only if the video orientation actually
changed, otherwise the continuous size re-transposition affects the visual rendering of the video
element.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateVideoOrientation):
(WebCore::MediaPlayerPrivateGStreamer::setVideoSourceOrientation):

Canonical link: <a href="https://commits.webkit.org/257912@main">https://commits.webkit.org/257912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa43704bc5ed035710d7d053edb578093c82eed2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109523 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169756 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104188 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10262 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92617 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107412 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91034 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34441 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89712 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22429 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77381 "Found 5 new API test failures: /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/js-exception, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/console-api, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/security-error, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/network-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3122 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23945 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3100 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9228 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43433 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5433 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4932 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3708,6 +3697,17 @@ void MediaPlayerPrivateGStreamer::setVideoSourceOrientation(ImageOrientation ori
+        m_player->sizeChanged();
+        return AbortableTaskQueue::Void();
+    }).has_value();
+    GST_DEBUG_OBJECT(pipeline(), "Finished waiting for main-thread task to call sizeChanged()... %s", sizeChangedProcessed ? "sizeChanged() was called." : "task queue aborted by flush");
```
setVideoSourceOrientation() is also called by updateVideoSizeAndOrientationFromCaps() who is called from a main-thread task in a lambda enqueued in videoSinkCapsChanged(). In that case this code would deadlock, as the main thread would be waiting for itself.

philn replied:
So if we're in main thread we shouldn't call `m_sinkTaskQueue.enqueueTaskAndWait`?

philn replied:
Also with this patch there's potentially 2 `sizeChanged` notifications in the `updateVideoSizeAndOrientationFromCaps()` code path... I'll rework this....

ntrrgc reviewed (changes requested):
I have identified a deadlock in the changes.

Our handling of video sizes is definitely buggy and I wouldn't be surprised if it needed a more through revamp.

philn committed:
[GStreamer] Video element keeps changing the aspect ratio randomly (when the orientation information is in video's metadata)
https://bugs.webkit.org/show_bug.cgi?id=246824
<rdar://problem/101646324>

Reviewed by Xabier Rodriguez-Calvar.

Emit sizeChanged notifications towards the parent player only if the video orientation actually
changed, otherwise the continuous size re-transposition affects the visual rendering of the video
element.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateVideoOrientation):
(WebCore::MediaPlayerPrivateGStreamer::setVideoSourceOrientation):

Canonical link: https://commits.webkit.org/257912@main

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3391,18 +3391,21 @@ static ImageOrientation getVideoOrientation(const GstTagList* tagList)
 {
     GST_DEBUG_OBJECT(pipeline(), "Updating orientation from %" GST_PTR_FORMAT, tagList);
-    setVideoSourceOrientation(getVideoOrientation(tagList));
+    auto sizeActuallyChanged = setVideoSourceOrientation(getVideoOrientation(tagList));
```
setVideoSourceOrientation() returns `true` if the orientation changed. Orientation could change without changing the size by doing any 180 degree rotation from its previous rotation.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3391,18 +3391,21 @@ static ImageOrientation getVideoOrientation(const GstTagList* tagList)
+        return;
 
     // If the video is tagged as rotated 90 or 270 degrees, swap width and height.
     if (m_videoSourceOrientation.usesWidthAsHeight())
         m_videoSize = m_videoSize.transposedSize();
```
This only works for the first rotation. For instance:

Case 1:
1. "rotate-90" is received, m_videoSize is transposed, getting vertical video (correct).
2. Default rotation is received, m_videoSize is left unaltered, resulting in playing horizontal video with vertical video dimensions.

Case 2:
1. "rotate-90" is received, m_videoSize is transposed, getting vertical video (correct).
2. "rotate-270" is received, m_videoSize is wrongly transposed again, resulting in playing vertical video as horizontal video.

Instead of doing mutable manipulations, it would be safer to keep two video sizes, one before the rotation and one after the rotation. This may look like adding complexity, but in reality, the complexity is already there, so it's more like being honest about it.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3391,18 +3391,21 @@ static ImageOrientation getVideoOrientation(const GstTagList* tagList)
 void MediaPlayerPrivateGStreamer::updateVideoOrientation(const GstTagList* tagList)
```
It would be nice to note in a comment that this function will be called from the streaming thread of the sink.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3711,15 +3714,16 @@ RefPtr<VideoFrame> MediaPlayerPrivateGStreamer::videoFrameForCurrentTime()
-        return;
+        return false;
 
     m_videoSourceOrientation = orientation;
```
All these data members (`m_videoSourceOrientation`, `m_videoSize`, `m_textureMapperFlags`) that are being read and written from the sink streaming thread should not be carelessly with the data members that are read and written from the main thread. Any interaction should be carefully considered, as the potential (and most likely, reality) of race conditions is great. 

Currently there is no documentation for what members of MediaPlayerPrivateGStreamer are meant to be read and written by the sink streaming thread and what members are meant to be used by the main thread, and the data members of the former are not even grouped together. I also have an impression that maybe some of those should be separated from MediaPlayerPrivateGStreamer and be part of the sinks.

ntrrgc reviewed (commented):
The rotation handling code is still in need of a revamp.
